src/sentry/utils/registry.py

```diff
@@ -0,0 +1,35 @@
+from __future__ import annotations
+
+from typing import Generic, TypeVar
+
+
+class AlreadyRegisteredError(ValueError):
+    pass
+
+
+class NoRegistrationExistsError(ValueError):
+    pass
+
+
+T = TypeVar("T")
+
+
+class Registry(Generic[T]):
+    def __init__(self):
+        self.registrations: dict[str, T] = {}
+
+    def register(self, key: str):
```

Comment:
🤔 should this class also have a way to register without a decorator? (although, it's probably better to not so we can standardize this further)

Reply:
We do have a few registries that don't use decorators - deletions and features come to mind.

One drawback to decorator based registries is that they rely on side-effects from imports. Should we also have a hook for registries to import a list of modules so that we don't have silent failures?

Reply:
If that case comes up I think we can add to it. Right now you can still use this to register things as well, it's just clunky.

`your_registry.register("some_name")(your_registered_thing)`

Reply:
> We do have a few registries that don't use decorators - deletions and features come to mind.
> 
> One drawback to decorator based registries is that they rely on side-effects from imports. Should we also have a hook for registries to import a list of modules so that we don't have silent failures?

Yeah, this is generally a concern... I think we do need something like that, although at the end of the day it does rely on the implementor to know that they need to do it, and to remember to do it.

I might think about good ways to do this as I use the registry. Because I'm not sure what a great solution is right now